src/graphene.cpp

```diff
@@ -0,0 +1,1471 @@
+                      "(re-requested tx), compression ratio %3.2f, peer=%s\n",
+            pfrom->grapheneBlock.GetHash().ToString(), blockSize, pfrom->nSizeGrapheneBlock, nSizeGrapheneBlockTx,
+            ((float)blockSize) / ((float)pfrom->nSizeGrapheneBlock + (float)nSizeGrapheneBlockTx), pfrom->GetLogName());
+
+        // Update run-time statistics of graphene block bandwidth savings.
+        // We add the original graphene block size with the size of transactions that were re-requested.
+        // This is NOT double counting since we never accounted for the original graphene block due to the re-request.
+        graphenedata.UpdateInBound(nSizeGrapheneBlockTx + pfrom->nSizeGrapheneBlock, blockSize);
+        LOG(GRAPHENE, "Graphene block stats: %s\n", graphenedata.ToString());
+
+        PV->HandleBlockMessage(pfrom, strCommand, MakeBlockRef(pfrom->grapheneBlock), inv);
+    }
+
+    return true;
+}
+
+CRequestGrapheneBlockTx::CRequestGrapheneBlockTx(uint256 blockHash, std::set<uint64_t> &setHashesToRequest)
+{
+    blockhash = blockHash;
+    setCheapHashesToRequest = setHashesToRequest;
+}
+
+bool CRequestGrapheneBlockTx::HandleMessage(CDataStream &vRecv, CNode *pfrom)
+{
+    if (!pfrom->GrapheneCapable())
+    {
+        dosMan.Misbehaving(pfrom, 100);
+        return error("get_grblocktx message received from a non GRAPHENE node, peer=%s", pfrom->GetLogName());
+    }
+
+    CRequestGrapheneBlockTx grapheneRequestBlockTx;
+    vRecv >> grapheneRequestBlockTx;
+
+    // Message consistency checking
+    if (grapheneRequestBlockTx.setCheapHashesToRequest.empty() || grapheneRequestBlockTx.blockhash.IsNull())
+    {
+        dosMan.Misbehaving(pfrom, 100);
+        return error("Incorrectly constructed get_grblocktx received.  Banning peer=%s", pfrom->GetLogName());
+    }
+
+    // We use MSG_TX here even though we refer to blockhash because we need to track
+    // how many grblocktx requests we make in case of DOS
+    CInv inv(MSG_TX, grapheneRequestBlockTx.blockhash);
+    LOG(GRAPHENE, "Received get_grblocktx for %s peer=%s\n", inv.hash.ToString(), pfrom->GetLogName());
+
+    // Check for Misbehaving and DOS
+    // If they make more than 20 requests in 10 minutes then disconnect them
+    {
+        LOCK(cs_vNodes);
+        if (pfrom->nGetGrapheneBlockTxLastTime <= 0)
+            pfrom->nGetGrapheneBlockTxLastTime = GetTime();
+        uint64_t nNow = GetTime();
+        pfrom->nGetGrapheneBlockTxCount *=
+            std::pow(1.0 - 1.0 / 600.0, (double)(nNow - pfrom->nGetGrapheneBlockTxLastTime));
+        pfrom->nGetGrapheneBlockTxLastTime = nNow;
+        pfrom->nGetGrapheneBlockTxCount += 1;
+        LOG(GRAPHENE, "nGetGrapheneTxCount is %f\n", pfrom->nGetGrapheneBlockTxCount);
+        if (pfrom->nGetGrapheneBlockTxCount >= 20)
+        {
+            dosMan.Misbehaving(pfrom, 100); // If they exceed the limit then disconnect them
```

Comment:
Only disconnect if mainnet.  We can easily exceed this on testnet, regtest, or nolnet due to how they are often lightly mined until somebody needs a bunch of coins.  We should also do the same in the other places were we limit these calls.

Reply:
What's the best way to test for this?

Reply:
You could run two nodes in regtest mode, connect them and generate > 20 blocks.  It should even be possible to write a python script to trigger this.

Reply:
Ah, what I meant was, from that place in the code, how can I detect that I'm on mainnet in order to know when to disconnect?